src/main/query/where.c

```diff
@@ -63,352 +63,265 @@ static int AerospikeQuery_Where_Add(AerospikeQuery *self, PyObject *py_ctx,
-                                        as_blob_equals(val, bytes_size, true));
-            }
-            else if (index_type == AS_INDEX_TYPE_LIST) {
-                as_query_where_with_ctx(
-                    &self->query, bin, pctx,
-                    as_blob_contains(LIST, val, bytes_size, true));
-            }
-            else if (index_type == AS_INDEX_TYPE_MAPKEYS) {
-                as_query_where_with_ctx(
-                    &self->query, bin, pctx,
-                    as_blob_contains(MAPKEYS, val, bytes_size, true));
-            }
-            else if (index_type == AS_INDEX_TYPE_MAPVALUES) {
-                as_query_where_with_ctx(
-                    &self->query, bin, pctx,
-                    as_blob_contains(MAPVALUES, val, bytes_size, true));
-            }
-            else {
-                rc = 1;
-                free(bin);
-                break;
-            }
-            if (py_ubin) {
-                Py_DECREF(py_ubin);
-                py_ubin = NULL;
-            }
+    // Can point to either val1_int or val1_str. We use this so we can pass in the same optional argument
+    // for both ints or strs to as_query_where_with_ctx().
+    void *val1 = NULL;
 
-            self->query.where.entries[0].value.blob_val._free = true;
+    Py_ssize_t bytes_size = 0;
 
-            // Cleanup
-            free(bin);
+    if (in_datatype == AS_INDEX_STRING || in_datatype == AS_INDEX_GEO2DSPHERE) {
+        if (!PyUnicode_Check(py_val1)) {
+            goto CLEANUP_ON_ERROR1;
         }
-        else {
-            // If it ain't expected, raise and error
-            as_error_update(
-                &err, AEROSPIKE_ERR_PARAM,
-                "predicate 'equals' expects a string or integer value.");
-            PyObject *py_err = NULL;
-            error_to_pyobject(&err, &py_err);
-            PyErr_SetObject(PyExc_Exception, py_err);
-            rc = 1;
-            break;
+        const char *buffer = PyUnicode_AsUTF8(py_val1);
+        if (!buffer) {
+            goto CLEANUP_ON_ERROR1;
         }
-
-        break;
+        val1_str = strdup(buffer);
+        val1 = (void *)val1_str;
     }
-    case AS_PREDICATE_RANGE: {
-        if (in_datatype == AS_INDEX_NUMERIC) {
-            if (PyUnicode_Check(py_bin)) {
-                py_ubin = PyUnicode_AsUTF8String(py_bin);
-                bin = PyBytes_AsString(py_ubin);
-            }
-            else if (PyByteArray_Check(py_bin)) {
-                bin = PyByteArray_AsString(py_bin);
-            }
-            else {
-                rc = 1;
-                break;
-            }
-            int64_t min = pyobject_to_int64(py_val1);
-            int64_t max = pyobject_to_int64(py_val2);
+    else if (in_datatype == AS_INDEX_NUMERIC) {
+        val1_int = pyobject_to_int64(py_val1);
+        if (PyErr_Occurred()) {
+            PyErr_Clear();
+            val1 = 0;
+        }
+        val1 = (void *)val1_int;
```

Comment:
Double checking that this assignment is correct. This is setting the value of the void pointer to val1_int. Shouldn't it be pointing to it instead, like `val1 = (void *)&val1_int` ?

Reply:
Yeah this is intentional. When the C client reads `val1` as an argument, in this case, it will type cast it as a signed 64-bit integer before getting the value.